Python/Product/Analysis/ProjectEntry.cs

```diff
@@ -84,86 +82,74 @@ IAnalysisCookie cookie
         public event EventHandler<EventArgs> OnNewParseTree;
         public event EventHandler<EventArgs> OnNewAnalysis;
 
-        public void UpdateTree(PythonAst newAst, IAnalysisCookie newCookie) {
-            lock (this) {
-                if (_updatesPending > 0) {
-                    _updatesPending--;
+        private readonly ManualResetEventSlim _pendingParse = new ManualResetEventSlim(true);
+        private long _expectedParse;
+
+        private class ActivePythonParse : IPythonParse {
+            private readonly ProjectEntry _entry;
+            private readonly long _expected;
+            private bool _completed;
```

Comment:
Has to be either volatile or accessed under lock

Reply:
Each instance of this class is only used on a single thread - it's the call back into ProjectEntry that needs to be controlled.